### ISSUE TITLE
`invalid-return-type` false positive on constrained generic when narrowing for each constraint

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#621]
### Summary

```py
def foo[T: (int, str)](value: T) -> T:
    if isinstance(value, int):
        return 1
    return ""
```
```
Return type does not match returned value: expected `T`, found `Literal[1]` (invalid-return-type) [Ln 3, Col 16]
Return type does not match returned value: expected `T`, found `Literal[""]` (invalid-return-type) [Ln 4, Col 12]
```
https://play.ty.dev/2aab612b-e197-46d2-a47f-7a288cbe3574

since this is a constraint instead of a regular generic, this should be allowed

**Comment by 90degs2infty** [comment]
Similar to the above:

```py
def or_default[T](value: T | None, default: T) -> T:
    return default if value is None else value


def or_empty(value: str | None) -> str:
    return or_default(value, "")
```

Triggers
```
Return type does not match returned value: expected `str`, found `str | None` (invalid-return-type) [Ln 6, Col 12]
```
but to my understanding `or_default` will always return a `T`.

https://play.ty.dev/5e93f1bf-8166-4f5d-b7a2-6e4fea38265a

**Comment by carljm** [comment]
The second comment here is not the same issue as the OP, it's a duplicate of #1115, which will be fixed by #623.
